### PR TITLE
Adding graphql endpoint

### PR DESCRIPTION
**Pull request by arnarthor**
### What is this?

Adding a graphql endpoint that in time will have the full schema of all objects available at apis.is exposed through this single schema.

##### Where does this stand?

Currently this has implemented a graphql schema for the `/address` and `/bus/realtime` endpoints with the exact same data structure as the REST endpoints. This schema is accessible at `/graphql`. It accepts POST requests with `query` as a body field. 

Example here
```
{
  "query": "{
    busses(busses: \"1\") {
      busNr 
      busses {
        x 
        y 
        unixTime
      }
    }
  }"
}
```

This would be the equivalent of doing a get request to `/bus/realtime?busses=1` except this only returns the specified fields (in this case, busNr and busses array with the x, y and unixTime fields).

Additionally this exposes a `graphiql` interface if accessed at http://apis.com/graphql in a browser. `graphiql` can be used to test, debug and inspect the schema. More information on that can be found [here](https://github.com/graphql/graphiql)

### Why?

The main reason I decided to do this was because we've shown a lot of interest in graphql at QuizUp. I wanted to investigate it further and thought this would be a great starting point since it's a proven API with lots of data that could be cool to expose as a graphql schema.

Having this as a single graphql schema could allow us to do cool things. An example I've thought of would be to do a single fetch that would get the tv schedule and the handball schedule by asking for both at the same time using a graphql query. You could then use that data to do a match to see what games are shown in direct TV.

The main benefit in the example above is that you wouldn't need to do two http requests for the data, and you could ask for specific fields like the name of the teams, time of game and then the tv schedule time and name of show. This data should be sufficient to do the match to see if a single game is shown directly on TV. No unnecessary data would be sent over the wire which could speed up queries on mobile devices and such.

### Checklist

- [ ] Write tests
- [ ] Write documentation

### CODE REVIEW

File: endpoints/address/index.js [hunk 1]
```diff
@@ -3,6 +3,34 @@ import h from 'apis-helpers'
 import app from '../../server'
 import _ from 'lodash'
 
+const lookupAddresses = (address) => {
+  return new Promise((resolve, reject) => {
```
Comment (tryggvigy) [inline endpoints/address/index.js]: Could be simplified a bit to skip one level of indentation 👼 
```js
const lookupAddresses = (address) => new Promise((resolve, reject) => {
  ...
}
```

Comment (arnarthor) [inline endpoints/address/index.js]: true dat

File: endpoints/address/index.js [hunk 2]
```diff
@@ -37,7 +25,25 @@ app.get('/address/:address?', (req, res) => {
...
+  if (address === '') {
+    res.status(400).json({
+      error: 'Please provide a valid address to lookup',
+    })
+  }
+
+  lookupAddresses(address).then(
+    (results) => res.cache().json({ results }),
+    () => res.status(500).json({ error: 'www.postur.is refuses to respond or give back data' })
+  )
```
Comment (tryggvigy) [inline endpoints/address/index.js]: Do we need a `.catch` block here ?

Comment (arnarthor) [inline endpoints/address/index.js]: No, the latter argument of `.then` is `onFailure`

File: endpoints/graphql/index.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import graphqlHTTP from 'express-graphql'
+import bodyParser from 'body-parser'
+import schema from './schemas'
+import app from '../../server'
+
+app.use('/graphql', bodyParser.json(), graphqlHTTP({
+  schema,
+  graphiql: true,
+}))
```
Comment (tryggvigy) [inline endpoints/graphql/index.js]: 💥 ❤️ 🙈  awesome

**Comment by koddsson** [comment]
Nice. I think I won't be making any code comments until this is out of WIP but could you eleborate on what problem graphql helps us with for those of us that haven't checked it out yet? ☺️ 

CC-ing @kristjanmik since I think he has been working with/is interested in graphql.

**Comment by arnarthor** [comment]
The main reason I decided to do this was because we've shown a lot of interest in graphql at QuizUp. I wanted to investigate it further and thought this would be a great starting point since it's a proven API with lots of data that could be cool to expose as a graphql schema.

Having this as a single graphql schema could allow us to do cool things. An example I've thought of would be to do a single fetch that would get the tv schedule and the handball schedule by asking for both at the same time using a graphql query. You could then use that data to do a match to see what games are shown in direct TV.

The main benefit in the example above is that you wouldn't need to do two http requests for the data, and you could ask for specific fields like the name of the teams, time of game and then the tv schedule time and name of show. This data should be sufficient to do the match to see if a single game is shown directly on TV. No unnecessary data would be sent over the wire which could speed up queries on mobile devices and such.

**Comment by kristjanmik** [comment]
Great work!
I agree on the file structure point, we should try to keep all endpoint specific code inside the endpoint folder(until a better solution is found). It might be best to merge this one early since it touches a lot of code and most of the GraphQL ground work has been laid out (by you). We could then add every endpoint as a separate pull request.

**Comment by arnarthor** [comment]
Sonuds good @kristjanmik! I'll create a PR towards the original repo

**Comment by arnarthor** [comment]
I've updated the description of the PR. If you feel this is missing something or have some comments on the code feel free to do so, I'll fix it ASAP.

**Comment by kristjanmik** [comment]
Super 👍 I will go ahead and merge this tomorrow unless someone has any objections to the prescribed procedure.

**Comment by tryggvigy** [comment]
I'm looking forward to this merge.
I think adding a GraphQL endpoint to apis.is amazing.
apis.is has been a great learning tool for Icelandic developers (made my first http request to apis.is ❤️ ) and providing a GraphQL server is very empowering to developers who are experimenting with GraphQL clients to their applications, students who want to try out react+relay or just anyone who wants to understand what GraphQL is all about.

I have already started experimenting on this PR locally and I'm more than happy to help in adding the rest of the endpoints to the GraphQL server.

kudos to @arnarthor for the great initiative 🎉

**Comment by johannth** [comment]
Have you taken a look at https://github.com/syrusakbary/gdom? The idea is to expose the HTML structure of a webpage as a graphQL query, allowing you extract structure from it using CSS selectors. It is written in Python and has limitations but it this approach could fit apis.is in simple cases. The approach of declaring the structure of the valuable data using a GraphQL query feels simple and could work in many cases without having to do more advanced data transformation.

You can play around with it here: http://gdom.graphene-python.org/. After five minutes I could extract some [structure](http://gdom.graphene-python.org/?query=%7B%0A%20%20page(url%3A%22http%3A%2F%2Fwww.elko.is%2Felko%2Fis%2Fvorur%2Futsala_-_%25C3%259Evottavelar%2FWhirlpool_thvottavel_AWOD7114.ecp%22)%20%7B%0A%20%20%20%20item%3A%20query(selector%3A%20%22.hProduct%22)%20%7B%0A%20%20%20%20%20%20title%3A%20text(selector%3A%20%22%23producttitle%22)%0A%20%20%20%20%20%20price%3A%20text(selector%3A%20%22.discount-price%22)%0A%20%20%20%20%20%20specs%3A%20query(selector%3A%20%22%23specs%20tr.product_property%22)%20%7B%0A%20%20%20%20%20%20%09key%3A%20text(selector%3A%20%22td.product_property_name%22)%0A%20%20%20%20%20%20%20%20value%3A%20text(selector%3A%20%22td.product_property_value%22)%0A%20%20%20%20%09%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D) from the Elko product page on washing machines. The biggest downside so far is that using the selectors I couldn't choose data on the text contents of a cell.

**Pull request merged**